src/transformers/models/bamba/modeling_bamba.py

```diff
@@ -942,10 +933,51 @@ def extra_repr(self):
         return f"{tuple(self.weight.shape)}, eps={self.variance_epsilon}"
 
 
+def get_seq_idx_from_position_ids(position_ids: torch.LongTensor) -> torch.Tensor:
+    # Needs to handle multiple cases. Elements in a batch may look like:
+    #
+    # 1) Packed sequences, which look a bunch of `torch.range`'s concatenated together, as in
+    # [0, 1, 2, 0, 1, 2, 3, ...]. Encountered during training.
+    #
+    # 2) Generation with a cache: position ids can start at arbitrary values, as in [2, 3, 4, ...]
+    #
+    # 3) Left-padded inputs during generation. prepare_inputs_for_generation encodes all masked-out
+    # positions as position_ids = 1, so this scenario presents as [1, 1, 1, 0, 1, 2, 3, ...]
+    device = position_ids.device
+
+    seq_idx_list = []
+    idxs = torch.arange(1, position_ids.shape[1], device=device, dtype=torch.int32)
+    for pos_ids_batch in position_ids:
+        non_increasing_pos_id = pos_ids_batch[1:] <= pos_ids_batch[:-1]
+        cu_seq_lens = torch.cat(
+            (
+                torch.tensor([0], device=device, dtype=torch.int32),
+                idxs[non_increasing_pos_id],
+                torch.tensor(pos_ids_batch.shape, device=device, dtype=torch.int32),
+            ),
+        )
```

Comment:
Using this `non_increasing_pos_id` indexing seems to cover all the `poisition_ids` cases we need to consider during training and generation. See any edge cases here @fabianlim ?